File: OsmAnd-java/src/main/java/net/osmand/router/VehicleRouter.java
package net.osmand.router;

import java.util.List;
import java.util.Map;

import net.osmand.binary.RouteDataObject;
import net.osmand.router.BinaryRoutePlanner.RouteSegment;
import net.osmand.shared.routing.GeneralRouterProfile;

public interface VehicleRouter {

	
	public boolean containsAttribute(String attribute);
	
	public String getAttribute(String attribute);
	
	/**
	 * return if the road is accepted for routing
	 */
	public boolean acceptLine(RouteDataObject way);
	
	/**
	 * return oneway +/- 1 if it is oneway and 0 if both ways
	 */
	public int isOneWay(RouteDataObject road);
	
	/**
	 * return penalty transition in seconds
	 */
	public float getPenaltyTransition(RouteDataObject road);
	
	/**
	 * return delay in seconds (0 no obstacles)
	 */
	public float defineObstacle(RouteDataObject road, int point, boolean isBackwardDir);
	
	/**
	 * return delay in seconds for height obstacles
	 */
	public double defineHeightObstacle(RouteDataObject road, short startIndex, short endIndex);
	
	/**
	 * return delay in seconds (0 no obstacles)
	 */
	public float defineRoutingObstacle(RouteDataObject road, int point, boolean isBackwardDir);

	/**
	 * return routing speed in m/s for vehicle for specified road
	 */
	public float defineRoutingSpeed(RouteDataObject road, boolean dir);
	
	/**
	 * return real speed in m/s for vehicle for specified road
	 */
	public float defineVehicleSpeed(RouteDataObject road, boolean dir);
	
	/**
	 * define priority to multiply the speed for g(x) A* 
	 */
	public float defineSpeedPriority(RouteDataObject road, boolean dir);
	
	/**
	 * define destination priority
	 */
	float defineDestinationPriority(RouteDataObject road);

	/**
	 * Used for A* routing to calculate g(x)
	 * 
	 * @return minimal speed at road in m/s
	 */
	public float getDefaultSpeed();

	/**
	 * Used as minimal threshold of default speed
	 *
	 * @return minimal speed at road in m/s
	 */
	public float getMinSpeed();

	/**
	 * Used for A* routing to predict h(x) : it should be great any g(x)
	 * 
	 * @return maximum speed to calculate shortest distance
	 */
	public float getMaxSpeed();
	
	/**
	 * aware of road restrictions
	 */
	public boolean restrictionsAware();
	
	/**
	 * @param obj
	 * @return if road supports area routing
	 */
	public boolean isArea(RouteDataObject obj);
	
	/**
	 * Calculate turn time 
	 */
	public double calculateTurnTime(RouteSegment segment, RouteSegment prev);

	/**
	 * Calculate time to leave one road for another (e.g. getting on or off a ferry)
	 * @param from road segment before the change in the direction of travel
	 * @param to road segment after the change in the direction of travel
	 */
	default double calculateRoadChangeTime(RouteSegment from, RouteSegment to) {
		return 0;
	}

	/**
	 * Calculate time of a stop at the start point of the segment coming from its parent segment
	 * (e.g. a ferry at a terminal on the way)
	 */
	default double calculateStopTime(RouteSegment segment) {
		return 0;
	}

	/**
	 * Correct estimated times of the calculated route segments which depend on the whole route
	 * (e.g. getting on and off a ferry is paid only for a whole crossing)
	 */
	default void updateSegmentTimes(List<RouteSegmentResult> result) {
	}

		
	public VehicleRouter build(Map<String, String> params);

	public GeneralRouterProfile getProfile();
	
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/routing/VehicleRouter.kt
package net.osmand.shared.routing

/**
 * What one vehicle profile makes of a road: whether it may be used at all, how fast it is taken,
 * and what the obstacles and turns along it cost.
 *
 * The planner asks these questions for every road it considers, so everything here is on the hot
 * path of a route calculation.
 */
interface VehicleRouter {

	fun containsAttribute(attribute: String): Boolean

	fun getAttribute(attribute: String): String?

	/** Whether the road is accepted for routing. */
	fun acceptLine(way: RouteDataObject): Boolean

	/** +/- 1 when the road is one way, 0 when it may be taken both ways. */
	fun isOneWay(road: RouteDataObject): Int

	/** Penalty for entering the road, in seconds. */
	fun getPenaltyTransition(road: RouteDataObject): Float

	/** Delay at the point, in seconds, 0 when there is no obstacle. */
	fun defineObstacle(road: RouteDataObject, point: Int, isBackwardDir: Boolean): Float

	/** Delay for the climbs and descents between the two points, in seconds. */
	fun defineHeightObstacle(road: RouteDataObject, startIndex: Short, endIndex: Short): Double

	/** Delay at the point that only routing accounts for, in seconds. */
	fun defineRoutingObstacle(road: RouteDataObject, point: Int, isBackwardDir: Boolean): Float

	/** Speed the router plans with on this road, in m/s. */
	fun defineRoutingSpeed(road: RouteDataObject, dir: Boolean): Float

	/** Speed the vehicle is expected to actually travel at, in m/s. */
	fun defineVehicleSpeed(road: RouteDataObject, dir: Boolean): Float

	/** Factor the speed is multiplied by for g(x) of the A* search. */
	fun defineSpeedPriority(road: RouteDataObject, dir: Boolean): Float

	fun defineDestinationPriority(road: RouteDataObject): Float

	/**
	 * Used for A* routing to calculate g(x).
	 * @return minimal speed at road in m/s
	 */
	fun getDefaultSpeed(): Float

	/**
	 * Used as minimal threshold of default speed.
	 * @return minimal speed at road in m/s
	 */
	fun getMinSpeed(): Float

	/**
	 * Used for A* routing to predict h(x) : it should be great any g(x).
	 * @return maximum speed to calculate shortest distance
	 */
	fun getMaxSpeed(): Float

	/** Whether the profile is aware of road restrictions. */
	fun restrictionsAware(): Boolean

	/** Whether the road supports area routing. */
	fun isArea(obj: RouteDataObject): Boolean

	/** Cost of turning out of [prev] into [segment], in seconds. */
	fun calculateTurnTime(segment: RoadTraversal, prev: RoadTraversal): Double

	/** The same profile with [params] applied. */
	fun build(params: Map<String, String>): VehicleRouter

	fun getProfile(): GeneralRouterProfile
}
